table/index_builder.cc

```diff
@@ -60,32 +61,60 @@ PartitionedIndexBuilder::PartitionedIndexBuilder(
     const BlockBasedTableOptions& table_opt)
     : IndexBuilder(comparator),
       index_block_builder_(table_opt.index_block_restart_interval),
-      table_opt_(table_opt) {
-  sub_index_builder_ = IndexBuilder::CreateIndexBuilder(sub_type_, comparator_,
-                                                        nullptr, table_opt_);
-}
+      sub_index_builder_(nullptr),
+      table_opt_(table_opt) {}
 
 PartitionedIndexBuilder::~PartitionedIndexBuilder() {
   delete sub_index_builder_;
 }
 
+void PartitionedIndexBuilder::MakeNewSubIndexBuilder() {
+  assert(sub_index_builder_ == nullptr);
+  sub_index_builder_ = new ShortenedIndexBuilder(
+      comparator_, table_opt_.index_block_restart_interval);
+  flush_policy.reset(FlushBlockBySizePolicyFactory::NewFlushBlockPolicy(
+      table_opt_.metadata_block_size, table_opt_.block_size_deviation,
+      sub_index_builder_->index_block_builder_));
+}
+
 void PartitionedIndexBuilder::AddIndexEntry(
     std::string* last_key_in_current_block,
     const Slice* first_key_in_next_block, const BlockHandle& block_handle) {
-  sub_index_builder_->AddIndexEntry(last_key_in_current_block,
-                                    first_key_in_next_block, block_handle);
-  num_indexes++;
+  // Note: to avoid two consecuitive flush in the same method call, we do not
+  // check flush policy when adding the last key
   if (UNLIKELY(first_key_in_next_block == nullptr)) {  // no more keys
-    entries_.push_back({std::string(*last_key_in_current_block),
-                        std::unique_ptr<IndexBuilder>(sub_index_builder_)});
+    if (sub_index_builder_ == nullptr) {
+      MakeNewSubIndexBuilder();
+    }
+    sub_index_builder_->AddIndexEntry(last_key_in_current_block,
+                                      first_key_in_next_block, block_handle);
+    sub_index_last_key_ = std::string(*last_key_in_current_block);
```

Comment:
Yeah but this looks more explicit to me. Since it is error-prone part of the code (it caused a bug before) I would rather to make it explicit.